cirq-google/cirq_google/workflow/quantum_executable.py

```diff
@@ -215,6 +215,19 @@ def __str__(self):
     def __repr__(self):
         return _compat.dataclass_repr(self, namespace='cirq_google')
 
+    def __hash__(self) -> int:
+        if self._hash is None:  # type: ignore
+            object.__setattr__(self, '_hash', hash(dataclasses.astuple(self)))
+        return self._hash  # type: ignore
```

Comment:
The `_hash` is set unconditionally in `__init__`; do we need to handle it being `None` here? Also, I'm curious about how things were working before since we were never accessing `self._hash`; is that the member that the dataclass library itself uses to store the hash, or were were computing the hash but then never using the value?

Reply:
> The `_hash` is set unconditionally in `__init__`; do we need to handle it being `None` here?

Yes, `pickle.load` constructs objects without calling their `__init__` so a from-pickle-restored object will have `_hash` set to None per `__getstate__`.  I have tweaked this a bit in b8b7ff587bbd6f07b18213eb03e843378198eec1 so that the hash expressions are defined exactly once in class code.

> Also, I'm curious about how things were working before since we were never accessing `self._hash`; is that the member that the dataclass library itself uses to store the hash, or were were computing the hash but then never using the value?

I tried to overwrite the `_hash` value of a QuantumExecutable instance in the `main` version, but it did not change `hash(obj)` result.  The `_hash` attribute was possibly used in earlier versions of dataclass, but is not used now; we were indeed not using the `_hash` value.